### PR TITLE
use match statement instead of linear array search for some decode perf gains

### PR DESCRIPTION
**Pull request by mriise**
👉🤤🌈✋😅😔🌹🤦😅✋🌷😓😬🌟😅🌟🌈😅😝🥺✋🤤😈💪😅✋💦✋🥺🏃😓😻✋😅🌈😴😓🤘🍀😳😅👉✋😅🤘🌈🌟😻😝😅😈😴🌟🌈👏😅🌈🤘🥺✋😅😈😴✋🥺✋😅🤤🥺✋😅👉✋😈😈✋🥺😅😈😴🌟😻😝🌈😅😈😓😅😳😓💪😅👉🤘😈😅😈😴🌟🌈😅🌟🌈😅😚🤘😻👏😅🌖🌕🌕😚😓🥺😅😩🤘💞🤷😅😎😔🤪😅🌷🤤😈💃😴😅💦🌈😅🍀😓😓🙃🤘⭐💪😅🌷🤤😈💃😴😅🌈😈🤤😈✋🌷✋😻😈🌈😅🤤🥺✋😅😠😔🌹🤗😻🌈😅💦🌈😅😠😎☺🌹🤘🌈☺😅🌈✋✋😅😴😈😈⭐🌈✨👌👌😝🌟😈😴🤘👉☺💃😓🌷👌🌷🥺🌟🌟🌈✋👌👉🤤🌈✋😔🌹🤦✋🌷😓😬🌟👌😈🥺✋✋👌👉✋😻💃😴😅😚😓🥺😅👉✋😻💃😴✋🌈😅✨🤔🌖🌕🌕🤤🍀🌈😓😅😴🌟😅😬🤘🤤😻😅😱😎

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -20,31 +20,29 @@ pub type Alphabet = [char; 256];
...
+				codepoint: c,
+				index: i,
+			})?;
 		}
 
 		Ok(output)
 	}
 
 	fn encode(input: &Vec<u8>) -> String {
+		// is size esimation really necessary?
```
Comment (Jorropo) [inline src/lib.rs]:
```suggestion
```
This allows to do a single allocation instead of growing.

File: src/lib.rs [hunk 2]
```diff
@@ -101,4 +90,18 @@ mod tests {
...
+			println!("i{}:{}, {:?}", i, c, Emoji::get_index(c));
+			assert!(i as u8 == Emoji::get_index(c).unwrap());
+		}
+	}
+
+	#[test]
+	fn juan() {
+		let hi = "hi juan!";
+		println!("{}: {}", hi, Emoji::encode(&hi.as_bytes().to_owned()));
+	}
```
Comment (Jorropo) [inline src/lib.rs]:
```suggestion
```
Demo left in ?

File: src/lib.rs [hunk 3]
```diff
@@ -59,40 +57,31 @@ pub trait Base {
...
-		'❓', '💎', '💸', '⬇', '😨', '🌚', '🦋', '😷', '🕺', '⚠', '🙅', '😟', '😵', '👎', '🤲',
-		'🤠', '🤧', '📌', '🔵', '💅', '🧐', '🐾', '🍒', '😗', '🤑', '🌊', '🤯', '🐷', '☎', '💧',
-		'😯', '💆', '👆', '🎤', '🙇', '🍑', '❄', '🌴', '💣', '🐸', '💌', '📍', '🥀', '🤢', '👅',
-		'💡', '💩', '👐', '📸', '👻', '🤐', '🤮', '🎼', '🥵', '🚩', '🍎', '🍊', '👼', '💍', '📣',
-		'🥂',
-	];
+macro_rules! gen_alphabet {
+	($name:ident, $alphabet:literal) => {
+		impl Base for $name {
+			const ALPHABET: [char; 256] = const_str::to_char_array!($alphabet);
```
Comment (Jorropo) [inline src/lib.rs]:
```suggestion
			const ALPHABET: Alphabet = const_str::to_char_array!($alphabet);
```

**Review by Jorropo — CHANGES_REQUESTED** [review]

File: src/lib.rs (hunk 2 above)
Comment (mriise) [reply src/lib.rs]: no. its a useful test.

**Pull request merged**